wandb/sdk/artifacts/storage_handlers/fsspec_file_handler.py

```diff
@@ -0,0 +1,146 @@
+"""Storage handler utilizing fsspec."""
+import os
+import time
+from typing import TYPE_CHECKING, Any, Optional, Sequence, Union
+from urllib.parse import ParseResult
+
+from wandb.errors.term import termlog
+from wandb.sdk.artifacts.artifact_file_cache import get_artifact_file_cache
+from wandb.sdk.artifacts.artifact_manifest_entry import ArtifactManifestEntry
+from wandb.sdk.artifacts.storage_handler import DEFAULT_MAX_OBJECTS, StorageHandler
+from wandb.sdk.lib import filesystem
+from wandb.sdk.lib.hashutil import B64MD5, md5_string
+from wandb.sdk.lib.paths import FilePathStr, StrPath, URIStr
+
+if TYPE_CHECKING:
+    import fsspec  # type: ignore
+
+    from wandb.sdk.artifacts.artifact import Artifact
+
+
+class FsspecFileHandler(StorageHandler):
+    """Handles a variety of different storage solutions."""
+
+    def __init__(self, scheme: Optional[str] = None) -> None:
+        """Track files or directories on a variety of filesystem.
+
+        For now, this handler supports oss (Alibaba Object Storage System)
+        A list of all available options can be found under the following link:
+        https://github.com/fsspec/filesystem_spec/blob/master/fsspec/registry.py
+        """
+        self._scheme = scheme or "oss"
+        self._cache = get_artifact_file_cache()
+
+    def can_handle(self, parsed_url: "ParseResult") -> bool:
+        return parsed_url.scheme in self._scheme
+
+    def load_path(
+        self,
+        manifest_entry: ArtifactManifestEntry,
+        local: bool = False,
+    ) -> Union[URIStr, FilePathStr]:
+        """Load the file in the specified artifact given its corresponding entry.
+
+        Download the referenced artifact; return the path to the caller.
+
+        Arguments:
+            manifest_entry (ArtifactManifestEntry): The index entry to load
+
+        Returns:
+            (os.PathLike): A path to the file represented by `index_entry`
+        """
+        if not local:
+            assert manifest_entry.ref is not None
+            return manifest_entry.ref
+
+        if manifest_entry.ref is None:
+            raise ValueError(f"Cannot add path with no ref: {manifest_entry.path}")
```

Comment:
Nit: when validating data coming from the user it's important to have nice error messages. For data coming from the system an `assert` is enough. If the assert fails, it means we have a bug somewhere.
```suggestion
        assert manifest_entry.ref is not None
```